### DISCUSSION TITLE
How to obtain the gradient matrix of the element, perform operations, and assemble the matrices?

### DISCUSSION THREAD

**Discussion by tttsss-01** [idaholab/moose#28243] [Q&A Modules: Solid mechanics]
I want to calculate the $\bar{\mathbb{C}}$ as mentioned in the literature equation 31 and save its components in postprocessors.
$\bar{\mathbb{C}}$ is the overall tangent modulus tensor, a four order tensor. 
I think I should follow equation 23 first, perform operations on the gradient matrices $\boldsymbol{B}$ and $\mathbb{C}$ of each element, and then assemble them to obtain $\boldsymbol{K}$ and $\boldsymbol{L}$.
I am uncertain about how to do the following:
- How to obtain the gradient matrix $\boldsymbol{B}$ for each element, and what methods can be used to call $\boldsymbol{B}$?
- How to assemble the results calculated for each element.
- Which class do these matrices belong to, and what existing methods do they have for operations?

Thank you very much for any assistance or guidance!

![a1](https://github.com/user-attachments/assets/42f242a0-71df-4c70-bb9f-7f5a5cc1e2e8)
![a2](https://github.com/user-attachments/assets/32e742c9-0418-4718-b3d3-77aef69cc186)

**Comment by GiudGiud** [comment]
Hello

we do not form B currently afaik,  but kernels can access _test and its gradient for the currently considered test function.
You can look at Assembly.h for all the finite element arrays (shape, test functions, gradients etc) you can have access to

> How to assemble the results calculated for each element.

you will want to unroll the loop essentially, so the residual is assembled in a moose-like manner for a given test function
(see the section of the tutorial on kernels)

**Reply by tttsss-01** [reply]
Thank you for your help. I reviewed how things are done in the kernel. I noticed that in `StressDivergenceTensors` and `TotalLagrangianStressDivergenceBase`, $\boldsymbol{B}^T \mathbb{C} \boldsymbol{B}$  is computed at each quadrature point. For example, in `TotalLagrangianStressDivergenceBase`, I think this function returns the $\boldsymbol{B}^T \mathbb{C} \boldsymbol{B}$ at the quadrature points:
```C
template <class G>
Real
TotalLagrangianStressDivergenceBase<G>::computeQpJacobianDisplacement(unsigned int alpha,
                                                                      unsigned int beta)
{
  // J_{alpha beta} = phi^alpha_{i, J} T_{iJkL} G^beta_{kL}
  return gradTest(alpha).doubleContraction(_dpk1[_qp] * gradTrial(beta));
}
```
Then, in `Kernel::computeJacobian` and `Kernel::computeOffDiagJacobian`, these values are assembled into K (the specified component).
```C
void
Kernel::computeJacobian()
{
  prepareMatrixTag(_assembly, _var.number(), _var.number());

  precalculateJacobian();
  for (_i = 0; _i < _test.size(); _i++)
    for (_j = 0; _j < _phi.size(); _j++)
      for (_qp = 0; _qp < _qrule->n_points(); _qp++)
        _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * computeQpJacobian(); 

  accumulateTaggedLocalMatrix();

  if (_has_diag_save_in && !_sys.computingScalingJacobian())
  {
    DenseVector<Number> diag = _assembly.getJacobianDiagonal(_local_ke);
    Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
    for (const auto & var : _diag_save_in)
      var->sys().solution().add_vector(diag, var->dofIndices());
  }
}
```
I'm not entirely clear on what `prepareMatrixTag` and `accumulateTaggedLocalMatrix` do specifically. However, since we have already computed and assembled K, I assume there should be a way to access this K through some tag.

Similarly, I need to compute $\boldsymbol L$ (compute $\boldsymbol{B}^T \mathbb{C}$ and assemble) . Is it feasible to perform such calculations within a postprocessor?

Additionally, I need to compute $K^{-1}$. Does there exist a method to compute the inverse of K?

**Reply by GiudGiud** [reply]
accumulateTaggedLocalMatrix moves values from the local_ke threaded matrix to the global shared-between-threads matrix

>  However, since we have already computed and assembled K, I assume there should be a way to access this K through some tag.

add a matrix tag to all the kernels that contribute to K then you can retrieve it using that tag
https://mooseframework.inl.gov/source/interfaces/TaggingInterface.html

**Reply by tttsss-01** [reply]
I'm still not entirely sure how MOOSE assembles the Jacobian and off-diagonal Jacobian values into a global Jacobian matrix and then computes its inverse. 
I guess that the overall Jacobian is a block matrix composed of component * component sub-matrices, each of size N * N. Is  this correct?
I guess that during the Newton iteration process, the inverse of the Jacobian ($K^{-1}$) should have been calculated. Is there a way to access it directly?
If we were to recompute $K^{-1}$, do you have any suggestions?

**Reply by GiudGiud** [reply]
It does not compute its inverse directly. It solves the numerical system for the Newton update equation

What do you mean by N here?

You can have petsc compute the inverse in a user object if you want. This will only work well for tiny problems.

**Reply by tttsss-01** [reply]
N is the number of nodes

**Reply by GiudGiud** [reply]
> I guess that the overall Jacobian is a block matrix composed of component * component sub-matrices, each of size N * N. Is this correct?

N should be the number of DoFs here, not the number of nodes. If you are using 1st order Lagrange then for a single variable that is the case. 
For the block-matrices, are you asking about the DoFs are ordered for multiple variables?

**Reply by tttsss-01** [reply]
Yes, this is what I wanted to ask.

**Reply by GiudGiud** [reply]
DoF ordering should be by default: rank/variable group/dof object/variable

this can be changed by the user so be careful coding for a given ordering
